src/toxcore/dht/server/hole_punching.rs

```diff
@@ -0,0 +1,433 @@
+            self.last_recv_ping_time.elapsed() <= nat_ping_req_interval * 2 {
+                let ip = match HolePunching::get_major_ip(&addrs, MAX_CLIENTS_PER_FRIEND / 2) {
+                    None => return Box::new(future::ok(())),
+                    Some(ip) => ip,
+                };
+
+                if self.last_punching_time.elapsed() > Duration::from_secs(RESET_PUNCH_INTERVAL) {
+                    self.num_punch_tries = 0;
+                    self.first_punching_index = 0;
+                    self.last_punching_index = 0;
+                }
+
+                let ports_to_try = HolePunching::get_nat_ports(&addrs, ip);
+
+                let res = self.punch(ports_to_try, ip, server, friend_pk);
+
+                self.last_punching_time = Instant::now();
+                self.is_punching_done = true;
+
+                res
+        } else {
+            // NatPingResponse is not responded, or hole punching was already done,
+            // or the elapsed time is too long since we received NatPingResponse,
+            // then we do nothing.
+            Box::new(future::ok(()))
+        }
+    }
+
+    // Calc most common IP and if number of most common IP exceeds need_num, return it.
+    // need_num is normally 4, 4 is half of maximum close nodes per friend.
+    // When half of clients have same IP with different port, we consider it as
+    // friend is behind NAT.
+    fn get_major_ip(addrs: &[SocketAddr], need_num: u32) -> Option<IpAddr> {
+        let mut occurrences = HashMap::new();
+
+        for addr in addrs {
+            *occurrences.entry(addr.ip()).or_insert(0) += 1;
+        }
+
+        occurrences.into_iter().max_by_key(|&(_, count)| count)
+            .and_then(|(common_ip, count)|
+                if count > need_num {
+                    Some(common_ip)
+                } else {
+                    None
+                }
+            )
+    }
+
+    // Get ports list of given IP
+    fn get_nat_ports(addrs: &[SocketAddr], ip: IpAddr) -> Vec<u16> {
+        addrs.iter()
+            .filter(|addr| addr.ip() == ip)
+            .map(|addr| addr.port())
+            .collect::<Vec<u16>>()
+    }
+
+    // punch using first_punching_index
+    fn first_hole_punching(&mut self, ports: Vec<u16>, ip: IpAddr, server: &Server, friend_pk: PublicKey) -> IoFuture<()> {
+        let num_ports = ports.len();
+        let first_punching_index = self.first_punching_index;
+        let ping_sender = (0..MAX_PORTS_TO_PUNCH)
+            .map(|i| {
+                let it = i + first_punching_index;
+                let sign: i16 = if it % 2 == 1 { -1 } else { 1 };
+                let delta = sign * (it / (2 * num_ports as u32)) as i16;
+                let index = (it as usize / 2) % num_ports;
+                let port = (ports[index] as i16 + delta) as u16;
+
+                server.send_ping_req(
+                    &PackedNode::new(false, SocketAddr::new(ip, port), &friend_pk)
```

Comment:
I don't like PackedNode::new(**false**, please add inline comment to tell what is this like that: `/* is something turned on */ false`